### PR TITLE
fix the inheritance graph

### LINKED ISSUE #111: Fix the inheritance graph
**Problem summary/missing feature:**  
Inheritance throughout the library is a mess with a mix of super() calls and direct class instantiations.
All classes need to be looked after to make sure the Class.__init__() code is replaced by the rightful super().__init__() call.

**Expected behavior:**  
The library needs to behave exactly the same.

**Work to do:**  
Go through all the classes of all the modules to update the (multiple) inheritance calls.

### PR DESCRIPTION
**Pull request by harshiniwho**
# Pull Request Template

## Description

Cleaned up all the inheritance calls to the super classes using super().__init__() instead of className.__init__()

Fixes #111 fix the inheritance graph

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?

- I ran the existing test cases over the changes to ensure nothing breaks. (python -m unittest discover -s tests). 
I do have some questions regarding the tests, will add in a separate comment.

**Test Configuration**:
* OS: Ubuntu
* OS version: 16.04
* Python version: 3.8.3
* Architecture:

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by harshiniwho** [comment]
@arnauddupuis I left 3 "# to fix: tests fail if super class constructor is removed", the tests were breaking with some super() calls. Leaving the stack trace, will look into it tomorrow myself but if you have a clue on what might be breaking let me know:

e.g at line 985, board_items.py using super() leads to these types of errors.

======================================================================
ERROR: test_tile (test_boardItem.TestBoard)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/harshini/open-source/pygamelib/tests/test_boardItem.py", line 251, in test_tile
    bi = pgl_board_items.Tile()
  File "/home/harshini/open-source/pygamelib/pygamelib/board_items.py", line 1919, in __init__
    super().__init__(**kwargs)
  File "/home/harshini/open-source/pygamelib/pygamelib/board_items.py", line 1555, in __init__
    super().__init__(**kwargs)
  File "/home/harshini/open-source/pygamelib/pygamelib/board_items.py", line 983, in __init__
    super().__init__(**kwargs)
  File "/home/harshini/open-source/pygamelib/pygamelib/board_items.py", line 453, in __init__
    self.update_sprite()
  File "/home/harshini/open-source/pygamelib/pygamelib/board_items.py", line 487, in update_sprite
    self._item_matrix[row].append(self.base_item_type(**self.__kwargs))
  File "/home/harshini/open-source/pygamelib/pygamelib/board_items.py", line 1875, in __init__
    super().__init__(**kwargs)
  File "/home/harshini/open-source/pygamelib/pygamelib/board_items.py", line 1555, in __init__
    super().__init__(**kwargs)
  File "/home/harshini/open-source/pygamelib/pygamelib/board_items.py", line 983, in __init__
    super().__init__(**kwargs)
  File "/home/harshini/open-source/pygamelib/pygamelib/board_items.py", line 335, in __init__
    self.__is_restorable = self.parent.restorable()
  File "/home/harshini/open-source/pygamelib/pygamelib/board_items.py", line 1640, in restorable
    return self.__is_restorable
AttributeError: 'Tile' object has no attribute '_GenericStructure__is_restorable'

**Comment by arnauddupuis** [comment]
First of all, thanks for your help!
I'll look into it tonight, there is potentially a weird inheritance graph in BoardComplexItem. The reason is both simple and potentially not a good one: to keep the item system compatibility from breaking with complex items (items that can occupy more than one cell on a board) I created a BoardItemComplexComponent. Tile inherit from BoardComplexItem and is composed of BoardItemComplexComponent. 
BoardItemComplexComponent is introspecting its parent to morph its properties (restorable, overlappable, etc.) according to its parent's properties. 
From the errors you posted, I would say that there is an initialization issue with GenericStructure. Tile is a bit special as it inherits from GenericStructure and BoardComplexItem. And its base item is a GenericStructureComplexComponent who itself inherits from GenericStructure and BoardItemComplexComponent. Maybe the issue lies here (and maybe we could think of a better implementation for Tile...).
I'll investigate tonight.

**Comment by arnauddupuis** [comment]
Ok I think I traced it down and I know why it happens only with Tile.
The stack is pretty long but it tells the whole story: Tile inheritates from GenericStructure and BoardComplexItem and the base item of Tile (the component that are going to constitute the Tile) is GenericStructureComplexComponent. The instanciation of GenericStructure goes well but then during the BoardComplexItem instanciation the update_sprite() method is called. And if you look at where the super().__init__() call is in GenericStructure, the internal variables are not set yet. So when the sub items calls the parent's method to morph, the methods (like restorable() for example) try to access an uninitialized member. 

I fixed it by putting the super().__init__() call line 1555 after the intialization of the the __is_* members. I haven't check for side effects in the construction stack though.

File: pygamelib/board_items.py [hunk 1]
```diff
@@ -1553,7 +1552,7 @@ def __init__(self, **kwargs):
             kwargs["model"] = "#"
         if "name" not in kwargs.keys():
             kwargs["name"] = "structure"
-        Immovable.__init__(self, **kwargs)
+        super().__init__(**kwargs)
```
Comment (arnauddupuis) [inline pygamelib/board_items.py:1555]: This is the place I'm referencing in my comment.

**Comment by harshiniwho** [comment]
Thanks that was helpful fixing 2 of them. The last one -  I have gone around in circles now, I have moved things around to see if it gets fixed but nothing seemed to work.

At, line 1255. The Character import seems to also work fine elsewhere, however I don't think Character init() gets called? Or it could be something simple I am overlooking.

======================================================================
ERROR: test_tools_function (test_game.TestBase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/harshini/open-source/pygamelib/tests/test_game.py", line 257, in test_tools_function
    self.assertIsNone(g.save_board(1, "test-pygamelib.engine.Game.lvl1.json"))
  File "/home/harshini/open-source/pygamelib/pygamelib/engine.py", line 2451, in save_board
    data["library"].append(Game._obj2ref(o))
  File "/home/harshini/open-source/pygamelib/pygamelib/engine.py", line 2531, in _obj2ref
    ref["hp"] = obj.hp
AttributeError: 'NPC' object has no attribute 'hp'

----------------------------------------------------------------------

**Comment by arnauddupuis** [comment]
It looks like it's the same type of issue. If NPC does not have a "hp" member it means that the Character contructor was never called (or hasn't returned yet).

I'll look into it (tonight ^^).

**Comment by arnauddupuis** [comment]
That one is weird. It really looks like the NPC object does not get its members from Character, but when I try manually it's all there...
I'm sorry I didn't have a lot of time today. I'll keep investigating tomorrow.

**Comment by arnauddupuis** [comment]
_**TL;DR**:_ The solution is to make sure that **all** base objects of the library inheritates from `object`.

Got it!
This one bothered me a lot, so I reproduced the inheritance pattern:
```python
class Base:
    def __init__(self, **test):
        print(f"Base created test={test}")


class ChildA(Base):
    def __init__(self, **test):
        print(f">>> ChildA test={test}")
        super().__init__(**test)
        print("<<< ChildA")


class ChildB:
    def __init__(self, **test):
        print(f">>> ChildB test={test}")
        print("<<< ChildB")


class Dual(ChildA, ChildB):
    def __init__(self, **test):
        print(f">>> Dual test={test}")
        super().__init__(**test)
        print("<<< Dual")


Dual(a="a", b="b")
```
And got the same problem:
```bash
>>> Dual test={'a': 'a', 'b': 'b'}
>>> ChildA test={'a': 'a', 'b': 'b'}
Base created test={'a': 'a', 'b': 'b'}
<<< ChildA
<<< Dual
```
I wonder if the issue is not the lack of comon ancestor (as the inheritance graph could look like a diamond pattern to super()). Anyway I solved the issue by making sure that all base objects inherits from `object`.

Here is a simplified version/solution of what is happening in board_items.py (it's Player instead of NPC because it has the same issue):
```python
class BoardItem(object):
    def __init__(self, **kwargs):
        print("--> BoardItem")
        super().__init__()
        print("<-- BoardItem")


class Movable(BoardItem):
    def __init__(self, **kwargs):
        print("--> Movable")
        super().__init__(**kwargs)
        print("<-- Movable")


class Character(object):
    def __init__(self, **kwargs):
        print("--> Character")
        super().__init__()
        print("<-- Character")


class Player(Movable, Character):
    def __init__(self, **kwargs):
        print(f"--> Player {Player.mro()}")
        super().__init__(**kwargs)
        print("<-- Player")


Player()
```
And this behaves as expected:
```bash
--> Player [<class '__main__.Player'>, <class '__main__.Movable'>, <class '__main__.BoardItem'>, <class '__main__.Character'>, <class 'object'>]
--> Movable
--> BoardItem
--> Character
<-- Character
<-- BoardItem
<-- Movable
<-- Player
```

**Comment by harshiniwho** [comment]
That is really interesting, I don't usually inherit Object ever either in my base classes. Makes sense since I tried reversing the order of inheritance in NPC and had errors the other way around.
Shall make this change!

**Comment by arnauddupuis** [comment]
Great, thanks!
If you want to discuss this issue or others do not hesitate to join the discord server (link in the README_HACKTOBERFEST.md file)

**Comment by harshiniwho** [comment]
Fixed things, thanks for the help! Will join the discord server!

**Review by arnauddupuis — APPROVED** [review]
It looks good to me! Thanks.
